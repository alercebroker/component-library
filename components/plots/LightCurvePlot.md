```vue
<template>
    <light-curve-plot
      :detections="detections"
      :nonDetections="nonDetections"
      type="difference">
    </light-curve-plot>
</template>
<script>

import axios from "axios";

export default {
  data() {
    return {
        detections: [],
        nonDetections: [],
        oid: "ZTF20aaophpu"
    }
  },
  mounted(){
    axios
    .post("https://ztf.alerce.online/get_detections", { oid: this.oid })
    .then(response => {
      this.detections = response.data.result.detections;
    });
    axios
    .post("https://ztf.alerce.online/get_non_detections", { oid: this.oid })
    .then(response => {
      this.nonDetections = response.data.result.non_detections;
    });
  }
}
</script>
```