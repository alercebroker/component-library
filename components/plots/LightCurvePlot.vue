<template>
  <div>
    <v-chart :options="chartOptions.options" autoresize @click="onClick" />
  </div>
</template>

<script>
import { Vue, Component, Prop, Watch, Emit } from 'nuxt-property-decorator'
import Echarts from 'vue-echarts'
import { jdToDate } from '../utils/AstroDates.js'
import { DifferenceLightCurveOptions } from './plot-types/difference'
import { ApparentLightCurveOptions } from './plot-types/apparent'
import { FoldedLightCurveOptions } from './plot-types/folded'
import { LightCurveOptions } from './plot-types/utils/light-curve-utils'
import 'echarts/lib/chart/scatter'
import 'echarts/lib/component/toolbox'
import 'echarts/lib/component/tooltip'
import 'echarts/lib/component/legend'
import 'echarts/lib/component/dataZoom'
import 'echarts/lib/chart/custom'

@Component({
  components: {
    'v-chart': Echarts
  }
})
export default class LightCurvePlot extends Vue {
  @Prop({ type: Array, default: () => [] }) detections;
  @Prop({ type: Array, default: () => [] }) nonDetections;
  @Prop({ type: Number, default: 0 }) period;
  @Prop({ type: String }) type;

  showHelp = false;
  chartOptions = new LightCurveOptions();

  mounted () {
    this.generateOptions()
  }

  @Emit('detectionClick')
  onClick (detection) {
    const date =
      jdToDate(detection.value[0])
        .toUTCString()
        .slice(0, -3) + 'UT'
    return {
      mjd: detection.value[0],
      date,
      index: this.detections.findIndex(x => x.mjd === detection.value[0])
    }
  }

  generateOptions (type, detections, nonDetections, period) {
    if (type === 'difference') {
      const differenceLightCurveOptions = new DifferenceLightCurveOptions(detections, nonDetections)
      differenceLightCurveOptions.getSeries()
      differenceLightCurveOptions.getLegend()
      this.chartOptions = differenceLightCurveOptions
    }
    if (type === 'apparent') {
      const apparentLightCurveOptions = new ApparentLightCurveOptions(detections, nonDetections)
      apparentLightCurveOptions.getSeries()
      apparentLightCurveOptions.getLegend()
      this.chartOptions = apparentLightCurveOptions
    }
    if (type === 'folded') {
      const foldedLightCurveOptions = new FoldedLightCurveOptions(detections, nonDetections, period)
      foldedLightCurveOptions.getSeries()
      foldedLightCurveOptions.getLegend()
      this.chartOptions = foldedLightCurveOptions
    }
  }

  @Watch('type')
  onTypeChange (newType) {
    this.generateOptions(newType, this.detections, this.nonDetections, this.period)
  }

  @Watch('detections')
  onDetectionsChange (newDetections) {
    this.generateOptions(this.type, newDetections, this.nonDetections, this.period)
  }

  @Watch('nonDetections')
  onNonDetectionsChange (newNonDetections) {
    this.generateOptions(this.type, this.detections, newNonDetections, this.period)
  }
}
</script>

<style css>
.echarts {
  margin: auto;
  width: 100%;
  height: 100%;
  min-height: 350px;
}
</style>

