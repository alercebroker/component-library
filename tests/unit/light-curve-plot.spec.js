import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import Vue from 'vue'
import '@babel/polyfill'
import LightCurvePlot from "@/lib-components/light-curve-plot.vue"
import 'jest-canvas-mock'
import Echarts from "echarts"
describe('light-curve-plot.vue', () => {
  const { detections } = require('./detections.js')
  const mountFunction = (options) => {
    return shallowMount(LightCurvePlot, {
      ...options
    })
  }
  let spy;
  beforeAll(() => {
    spy = jest.spyOn(Echarts, 'getInstanceByDom').mockImplementation(() => {
      return {
        hideLoading: jest.fn(),
        setOption: jest.fn(),
        showLoading: jest.fn()
      }
    })
  })
  afterAll(() => {
    spy.mockRestore()
  })

  it('mounts echarts', () => {
    const wrapper = mountFunction({
      propsData: {
        detections: detections,
        type: "apparent"
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('sets chart options on mount', async () => {
    let generateOptions = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        detections: detections,
        type: "apparent"
      },
      methods: {
        generateOptions
      }
    })
    await Vue.nextTick();
    expect(generateOptions).toBeCalled()
  })

  it('sets right options for difference', async () => {
    const wrapper = mountFunction({
      propsData: {
        detections: detections,
        type: "difference"
      }
    })
    await Vue.nextTick()
    expect(wrapper.vm.scatter).toHaveProperty("legend")
    expect(wrapper.vm.scatter.legend.data).toStrictEqual(["g", "r", "g non-detections", "r non-detections"])
  })
  it('sets right options for apparent', async () => {
    const wrapper = mountFunction({
      propsData: {
        detections: detections,
        type: "apparent"
      }
    })
    await Vue.nextTick()
    expect(wrapper.vm.scatter).toHaveProperty("legend")
    expect(wrapper.vm.scatter.legend.data).toStrictEqual(["g", "r"])
  })

  it('sets right options for folded', async () => {
    const wrapper = mountFunction({
      propsData: {
        detections: detections,
        type: "folded"
      }
    })
    await Vue.nextTick()
    expect(wrapper.vm.scatter).toHaveProperty("xAxis")
    expect(wrapper.vm.scatter.xAxis.name).toStrictEqual("Phase")
  })

  it('emits detection selected with mjd', async () => {
    const wrapper = mountFunction({
      propsData: {
        detections: detections,
        type: "apparent"
      }
    })
    wrapper.vm.$emit('detectionClick', {
      mjd: detections[1].mjd,
      index: detections.findIndex(x => x.mjd == detections[1].mjd)
    })

    expect(wrapper.emitted('detectionClick')).toBeTruthy()
    expect(wrapper.emitted('detectionClick').length).toBe(1)
    expect(wrapper.emitted('detectionClick')[0][0].mjd).toEqual(detections[1].mjd)
    expect(wrapper.emitted('detectionClick')[0][0].index).toEqual(1)
  }) 
})