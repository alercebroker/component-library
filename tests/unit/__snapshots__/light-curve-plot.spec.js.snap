// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`light-curve-plot.vue mounts echarts 1`] = `
<div>
  <v-chart-stub options="[object Object]" autoresize="true"></v-chart-stub>
</div>
`;
